sql/src/main/java/io/crate/analyze/relations/RelationNormalizer.java

```diff
@@ -0,0 +1,337 @@
+                @Override
+                public Path apply(Field input) {
+                    return input.path();
+                }
+            });
+        }
+    }
+
+    private static class SymbolNormalizer extends SymbolVisitor<Void, Symbol> {
+
+        private static final SymbolNormalizer INSTANCE = new SymbolNormalizer();
+
+        public static void normalize(QuerySpec querySpec) {
+            querySpec.outputs(normalize(querySpec.outputs()));
+
+            if (querySpec.where().hasQuery() && !querySpec.where().noMatch()) {
+                Symbol query = normalize(querySpec.where().query());
+                querySpec.where(new WhereClause(query));
+            }
+
+            if (querySpec.orderBy().isPresent()) {
+                OrderBy orderBy = querySpec.orderBy().get();
+                List<Symbol> orderBySymbols = normalize(orderBy.orderBySymbols());
+                querySpec.orderBy(new OrderBy(orderBySymbols, orderBy.reverseFlags(), orderBy.nullsFirst()));
+            }
+
+            if (querySpec.groupBy().isPresent()) {
+                List<Symbol> groupBy = normalize(querySpec.groupBy().get());
+                querySpec.groupBy(groupBy);
+            }
+
+            if (querySpec.having().isPresent() && !querySpec.having().get().noMatch()) {
+                Symbol query = normalize(querySpec.having().get().query());
+                querySpec.having(new HavingClause(query));
+            }
+        }
+
+        private static Symbol normalize(Symbol symbol) {
+            return INSTANCE.process(symbol, null);
+        }
+
+        private static List<Symbol> normalize(Iterable<? extends Symbol> symbols) {
+            List<Symbol> outputs = new ArrayList<>();
+            for (Symbol symbol : symbols) {
+                outputs.add(normalize(symbol));
+            }
+            return outputs;
+        }
+
+        @Override
+        protected Symbol visitSymbol(Symbol symbol, Void context) {
+            return symbol;
+        }
+
+        @Override
+        public Symbol visitFunction(Function function, Void context) {
+            for (int i = 0; i < function.arguments().size(); i++) {
+                Symbol symbol = function.arguments().get(i);
+                function.setArgument(i, process(symbol, context));
+            }
+            return function;
+        }
+
+        @Override
+        public Symbol visitField(Field field, Void context) {
+            Field relationField = field.relation().getField(field.path(), Operation.READ);
+            if (relationField == null) {
+                return field;
+            }
+
+            Symbol output = ((QueriedRelation) field.relation()).querySpec().outputs().get(relationField.index());
+            return process(output, context);
+        }
+    }
+
+    private static class AggregatedReferenceVisitor extends SymbolVisitor<SymbolReferenceContext, Symbol> {
```

Comment:
you should probably use https://github.com/crate/crate/blob/6e5ba7c4b84604d9a4bd195de9fe3b5fcd79e0d3/sql/src/main/java/io/crate/analyze/symbol/SymbolVisitors.java#L40-L40